node/src/components/chain_synchronizer/operations.rs

```diff
@@ -1074,71 +1075,182 @@ async fn fetch_to_genesis(trusted_block: &Block, ctx: &ChainSyncContext<'_>) ->
+        if lowest_trusted_block_header.height() == 0 {
+            break;
+        }
+    }
+    Ok(())
+}
+
+// Fetches a batch of block headers, validates and stores in storage.
+// Returns either an error or lowest valid block in the chain.
+async fn fetch_block_headers_batch(
+    lowest_trusted_block_header: &BlockHeader,
+    ctx: &ChainSyncContext<'_>,
+) -> Result<BlockHeader, FetchBlockHeadersBatchError> {
+    let batch_id =
+        BlockHeadersBatchId::from_known(lowest_trusted_block_header, MAX_HEADERS_BATCH_SIZE);
+
+    loop {
+        let fetched_headers_data: FetchedData<BlockHeadersBatch> =
+            fetch_retry_forever::<BlockHeadersBatch>(ctx, batch_id).await?;
+        match fetched_headers_data {
+            FetchedData::FromStorage { item } => {
+                return BlockHeadersBatch::validate(
+                    &*item,
+                    lowest_trusted_block_header,
+                    ctx.config.verifiable_chunked_hash_activation(),
+                )
+                .map_err(FetchBlockHeadersBatchError::InvalidBatchFromStorage)
+            }
+            FetchedData::FromPeer { item, peer } => {
+                match BlockHeadersBatch::validate(
+                    &*item,
+                    lowest_trusted_block_header,
+                    ctx.config.verifiable_chunked_hash_activation(),
+                ) {
+                    Ok(new_lowest) => {
+                        info!(?batch_id, ?peer, "received valid batch of headers");
+                        ctx.effect_builder
+                            .put_block_headers_batch_to_storage(item.into_inner())
+                            .await;
+                        return Ok(new_lowest);
+                    }
+                    Err(err) => {
+                        error!(
+                            ?err,
+                            ?peer,
+                            ?batch_id,
+                            "block headers batch failed validation. Trying next peer..."
+                        );
+                        ctx.effect_builder.announce_disconnect_from_peer(peer).await;
+                    }
+                }
+            }
+        }
+    }
+}
+
+// Fetches blocks, their transactions and tries from Genesis until `trusted_block`.
+async fn fetch_blocks_and_state_since_genesis(ctx: &ChainSyncContext<'_>) -> Result<(), Error> {
+    info!("syncing blocks and deploys and state since Genesis");
+
+    let queue = Arc::new(WorkQueue::default());
+
+    // Flag set by a worker when it encounters an error.
+    let abort = Arc::new(AtomicBool::new(false));
+
+    for block_height in 0..=ctx.config.max_parallel_block_fetches() as u64 {
+        let block_hash = match ctx
+            .effect_builder
+            .get_block_hash_by_height_from_storage(block_height)
+            .await
+        {
+            Some(hash) => hash,
+            None => return Err(Error::NoSuchBlockHeight(block_height)),
```

Comment:
If this is error is only used here, the error name and message and such could reflect the fact that we expected the block to be there, since its header was fetched.

Reply:
You're right. I was thinking that I accomodate for any future use of the error but I should focus on the task at hand and expect devs coming after me to update the error message if necessary.